Adds support for Cholla's new Concatenation Format

mabruzzo opened the pull request:
## PR Summary

This PR adds support for loading Cholla datasets that are:
- distributed over multiple files
- stored in Cholla's historical concatenation datasets

For context:
- when Cholla writes out snapshots, it writes each block to a separate file.
- The Cholla repository provide python scripts to concatenate data, which has historically concatenate all of the data so that the resulting file would appear like there was only 1 process that evolved a single gigantic block. These scripts were commonly used to simplify data analysis.
- Until this PR, the yt-frontend has **only** known how to load datasets after they were concatenated in this manner. This introduced challenges because:
  1. the yt-frontend must load everything at once (introducing obvious memory issues)
  2. in cases where the number of cells exceeds 2^31 yt would simply break (this came up on large-memory nodes)

This PR introduces support for reading Cholla datasets that are either:
- distributed (i.e. they are not concatenated), OR
- use Cholla's newly added concatenation format.
  - I introduced the new concatenation format with the intent of improving performance when using yt (cholla-hydro/cholla#427). Essentially the new format organizes data in a manner based on the underlying block structure (in a similar manner to what an AMR simulation might do). This let's us reuse a lot of the machinery for reading distributed datasets
  - It's important that the yt-frontend can read in this format because (i) it provides some performance advantages (on parallel filesystems) and (ii) we have scripts to repackage datasets using the old concatenation format so that they now have this new format.

Aside: This PR superseded PR #4702 

## PR Checklist

- [x] New features are documented, with docstrings and narrative docs

mabruzzo committed:
Add support for distributed cholla datasets.

mabruzzo committed:
Update yt/frontends/cholla/data_structures.py

Co-authored-by: Matthew Turk <matthewturk@gmail.com>

mabruzzo committed:
Merge branch 'main' into cholla-frontend-improvements

mabruzzo committed:
improve the docstring for _split_fname_procid_suffix

mabruzzo committed:
Merge branch 'main' into cholla-frontend-improvements

mabruzzo committed:
add support for new cholla concatenation format

mabruzzo committed:
Merge branch 'main' into cholla-frontend-newformat

mabruzzo committed:
try to satisfy mypy

mabruzzo committed:
fix the cholla frontend

mabruzzo committed:
finally addressing the mypy type-checking issues

neutrinoceros commented on yt/frontends/cholla/data_structures.py:
```diff
@@ -9,16 +13,193 @@
+    # that are necessary to represent the process id. For flexibility, we will
+    # allow extra zero-padding
+
+    sep_i = filename.rfind(".")
```
I think this function could be much more readable if it was written around `str.rsplit` instead of `str.rfind`

neutrinoceros replied:
or better yet: `str.rpartition`

mabruzzo committed:
Update yt/frontends/cholla/data_structures.py

Co-authored-by: Clément Robert <cr52@protonmail.com>

mabruzzo committed:
address PR comments

mabruzzo commented on yt/frontends/cholla/data_structures.py:
```diff
@@ -9,16 +13,193 @@
+    # maps blockid to an index that select all associated data from a field-dataset
+    field_idx_map: Mapping[int, tuple[int | slice, ...]]
+
+    # in the future, we will add particle_group and particle_idx_map
```
fixed

mabruzzo committed:
Update yt/frontends/cholla/io.py

Co-authored-by: Clément Robert <cr52@protonmail.com>

mabruzzo committed:
fix a minor bug

mabruzzo committed:
Merge branch 'main' into cholla-frontend-newformat

mabruzzo committed:
Light refactoring

I realized that _split_fname_procid_suffix was no longer used for
anything. I think it's nice to have a function that explicitly describes
the relationship between the filename passed into yt.load and the
filename template. Consequently I replaced this function with a new
function called `_infer_fname_template` that we call from
`_determine_data_layout` (the contents didn't change dramatically)

mabruzzo commented:
@neutrinoceros, @matthewturk -- I just wanted to ping you to see if there is anything I can do to make this PR easier to review. I would love to see this get merged in time for the 4.5 release (whenever that is)

mabruzzo committed:
Merge branch 'main' into cholla-frontend-newformat

mabruzzo committed:
Introduced new tests for the Cholla frontend

-> this testing strategy creates synthetic data and performs a
   round-trip test to confirm that the data is loaded properly

mabruzzo committed:
Slightly reorganize Cholla-frontend

This is in anticipation of a followup PR adding support for particles

mabruzzo committed:
satisfy type-checker

mabruzzo committed:
satisfy type-check v2

mabruzzo committed:
Rename test_load.py and exclude it from nose

mabruzzo committed:
Merge branch 'main' into cholla-frontend-newformat

matthewturk commented:
Should we merge this?

mabruzzo commented:
That would be great!

mabruzzo commented:
@matthewturk and @neutrinoceros, I just wanted to ping you guys in case this fell off the todo list.

neutrinoceros commented:
Sorry, this is actually not part of my TODO list. I'm way behind on my backlog and I cannot predict how long it'll take me to give this the attention it needs
